internal/log/headers.go

```diff
@@ -0,0 +1,56 @@
+package log
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/rs/zerolog"
+
+	"github.com/pomerium/pomerium/internal/sets"
+)
+
+const (
+	headersFieldName   = "headers"
+	headersFieldPrefix = headersFieldName + "."
+)
+
+// GetHeaderField returns the header name for a field that represents logging a header value.
+func GetHeaderField[TField interface{ ~string }](field TField) (headerName string, ok bool) {
+	if strings.HasPrefix(string(field), headersFieldPrefix) {
+		return string(field)[len(headersFieldPrefix):], true
+	}
+
+	return "", false
+}
+
+// HTTPHeaders logs http headers based on supplied fields and a map of all headers.
+func HTTPHeaders[TField interface{ ~string }](
+	evt *zerolog.Event,
+	fields []TField,
+	src map[string]string,
+) *zerolog.Event {
+	all := false
+	include := sets.NewHash[string]()
+	for _, field := range fields {
+		if field == headersFieldName {
+			all = true
+			break
+		} else if strings.HasPrefix(string(field), headersFieldPrefix) {
+			include.Add(http.CanonicalHeaderKey(string(field[len(headersFieldPrefix):])))
```

Comment:
Can you help me understand: is this `include` set needed for the access logs, or the authorize logs, or both?

(I'm a little confused about what Envoy's [`request_headers`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/data/accesslog/v3/accesslog.proto#envoy-v3-api-field-data-accesslog-v3-httprequestproperties-request-headers) map contains in the absence of the `AdditionalRequestHeadersToLog` setting.)

Reply:
It is for access logs when called from the envoy control plane code and for authorize logs when called from the authorize service.

Envoy's `request_headers` contains nothing when `AdditionalRequestHeadersToLog` is empty.

Reply:
Maybe I'm still missing something... if we already configure the set of request headers from Envoy, then we wouldn't need to filter them out here, right? Does that mean this is only needed for the authorize logs, where we always start with the full set of request headers?

Reply:
Ok yes I can write different logic for the access logs.